source/beamline/BioXAS/BioXASBiStateGroup.cpp

```diff
@@ -0,0 +1,363 @@
+	}
+}
+
+bool BioXASBiStateGroup::addBiStateControl(AMControl *control, double state1Value, double state2Value)
+{
+	bool result = false;
+
+	if (control) {
+		controlState1ValueMap_.insert(control, state1Value);
+		controlState2ValueMap_.insert(control, state2Value);
+
+		addChildControl(control);
+
+		result = true;
+	}
+
+	return result;
+}
+
+bool BioXASBiStateGroup::removeBiStateControl(AMControl *control)
+{
+	bool result = false;
+
+	if ( control && (controlState1ValueMap_.contains(control) || controlState2ValueMap_.contains(control)) ) {
+		controlState1ValueMap_.remove(control);
+		controlState2ValueMap_.remove(control);
+
+		removeChildControl(control);
+
+		result = true;
+	}
+
+	return result;
+}
+
+bool BioXASBiStateGroup::clearBiStateControls()
+{
+	bool result = false;
+
+	QList<AMControl*> children = childControls();
+
+	if (children.count() > 0) {
+		foreach (AMControl *child, childControls())
+			removeBiStateControl(child);
+
+		result = true;
+	}
+
+	return result;
+}
+
+bool BioXASBiStateGroup::areChildrenState1() const
+{
+	bool result = false;
+
+	// Iterate through list of children, finding out if
+	// all are in state 1.
+
+	QList<AMControl*> children = childControls();
+
+	if (children.count() > 0) {
+		bool stateOK = true;
+
+		for (int i = 0, count = children.count(); i < count && stateOK; i++) { // we want to stop if any one child isn't ok.
+			AMControl *child = children.at(i);
+
+			if (child && isChildState1(child))
+				stateOK = true;
+			else
+				stateOK = false;
+		}
+
+		result = stateOK;
+	}
+
+	return result;
+}
+
+bool BioXASBiStateGroup::areChildrenState2() const
+{
+	bool result = false;
+
+	// Iterate through list of children, finding out if
+	// all are are state 2.
+
+	QList<AMControl*> children = childControls();
+
+	if (children.count() > 0) {
+		bool stateOK = true;
+
+		for (int i = 0, count = children.count(); i < count && stateOK; i++) { // we want to stop if any one child isn't ok.
+			AMControl *child = children.at(i);
+
+			if (child && isChildState2(child))
+				stateOK = true;
+			else
+				stateOK = false;
+		}
+
+		result = stateOK;
+	}
+
+	return result;
+}
+
+bool BioXASBiStateGroup::isChildState1(AMControl *child) const
+{
+	bool result = false;
+
+	if (child && controlState1ValueMap_.keys().contains(child)) {
+		if (child->value() == controlState1ValueMap_.value(child))
+			result = true;
+	}
+
+	return result;
+}
+
+bool BioXASBiStateGroup::isChildState2(AMControl *child) const
+{
+	bool result = false;
+
+	if (child && controlState2ValueMap_.keys().contains(child)) {
+		if (child->value() == controlState2ValueMap_.value(child))
+			result = true;
+	}
+
+	return result;
+}
+
+AMAction3* BioXASBiStateGroup::createMoveChildrenToState1Action()
+{
+	AMListAction3 *moveChildren = new AMListAction3(new AMListActionInfo3("Moving bistate group.", "Moving bistate group."), AMListAction3::Parallel);
+
+	foreach (AMControl *child, children_) {
+		AMAction3 *action = 0;
+
+		if (child) {
```

Comment:
Since this is a BioXAS class it is probably fine, but is it valid to assume that if you can't move that you just wait for it to go to a certain value?  There is probably an assumption made somewhere that makes sense for this situation, but I haven't noticed it yet.